File: Samples/Oxygene/Sample Server/MainForm.pas
namespace SampleServer;

interface

uses
  System.Windows.Forms,
  System.Drawing,
  RemObjects.InternetPack.StandardServers,
  RemObjects.InternetPack.Http,
  System.IO;
type
  MainForm = class(System.Windows.Forms.Form)
  {$REGION Windows Form Designer generated fields}
  private
    components: System.ComponentModel.Container := nil;
    method InitializeComponent;
  assembly
      pictureBox1: System.Windows.Forms.PictureBox;
      Label2: System.Windows.Forms.Label;
      txtLog: System.Windows.Forms.TextBox;
      btnAction: System.Windows.Forms.Button;
      lblLink: System.Windows.Forms.LinkLabel;
      lblPort: System.Windows.Forms.Label;
      nudPort: System.Windows.Forms.NumericUpDown;
      lblRoot: System.Windows.Forms.Label;
      txtRoot: System.Windows.Forms.TextBox;
      txtServerName: System.Windows.Forms.TextBox;
      lblServerName: System.Windows.Forms.Label;
      lblCount: System.Windows.Forms.Label;
      nudCount: System.Windows.Forms.NumericUpDown;
      lblUrl: System.Windows.Forms.Label;
    GroupBox1: System.Windows.Forms.GroupBox;
  {$ENDREGION}
  private
    fEchoServer: EchoServer;
    fHttpServer: SimpleHttpServer;
    method nudPort_ValueChanged(sender: System.Object; e: System.EventArgs);
    method MainForm_Closed(sender: System.Object; e: System.EventArgs);
    method MainForm_Load(sender: System.Object; e: System.EventArgs);
    method lblLink_LinkClicked(sender: System.Object; e: System.Windows.Forms.LinkLabelLinkClickedEventArgs);
    method ActivateServers();
    method DeactivateServers();
    method SetEnable(mode: Boolean);
    method AddLog(line: String);
    method LogRequest(aSender: object; ea: OnHttpRequestArgs);
  method btnAction_Click(sender: System.Object; e: System.EventArgs);
  protected
    method Dispose(aDisposing: boolean); override;
  public
    constructor;
    class method Main;
  end;


  LogRequestDelegate = delegate(request: String);


implementation

{$REGION Construction and Disposition}
constructor MainForm;
begin
  //
  // Required for Windows Form Designer support
  //
  InitializeComponent();

  //
  // TODO: Add any constructor code after InitializeComponent call
  //
end;

method MainForm.Dispose(aDisposing: boolean);
begin
  if aDisposing then begin
    if assigned(components) then
      components.Dispose();

    //
    // TODO: Add custom disposition code here
    //
  end;
  inherited Dispose(aDisposing);
end;
{$ENDREGION}

{$REGION Windows Form Designer generated code}
method MainForm.InitializeComponent;
begin
  var resources: System.ComponentModel.ComponentResourceManager := new System.ComponentModel.ComponentResourceManager(typeOf(MainForm));
  self.GroupBox1 := new System.Windows.Forms.GroupBox();
  self.lblUrl := new System.Windows.Forms.Label();
  self.nudCount := new System.Windows.Forms.NumericUpDown();
  self.lblCount := new System.Windows.Forms.Label();
  self.lblServerName := new System.Windows.Forms.Label();
  self.txtServerName := new System.Windows.Forms.TextBox();
  self.txtRoot := new System.Windows.Forms.TextBox();
  self.lblRoot := new System.Windows.Forms.Label();
  self.nudPort := new System.Windows.Forms.NumericUpDown();
  self.lblPort := new System.Windows.Forms.Label();
  self.lblLink := new System.Windows.Forms.LinkLabel();
  self.btnAction := new System.Windows.Forms.Button();
  self.txtLog := new System.Windows.Forms.TextBox();
  self.Label2 := new System.Windows.Forms.Label();
  self.pictureBox1 := new System.Windows.Forms.PictureBox();
  self.GroupBox1.SuspendLayout();
  (self.nudCount as System.ComponentModel.ISupportInitialize).BeginInit();
  (self.nudPort as System.ComponentModel.ISupportInitialize).BeginInit();
  (self.pictureBox1 as System.ComponentModel.ISupportInitialize).BeginInit();
  self.SuspendLayout();
  // 
  // GroupBox1
  // 
  self.GroupBox1.Anchor := (((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Left) 
        or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.GroupBox1.Controls.Add(self.lblUrl);
  self.GroupBox1.Controls.Add(self.nudCount);
  self.GroupBox1.Controls.Add(self.lblCount);
  self.GroupBox1.Controls.Add(self.lblServerName);
  self.GroupBox1.Controls.Add(self.txtServerName);
  self.GroupBox1.Controls.Add(self.txtRoot);
  self.GroupBox1.Controls.Add(self.lblRoot);
  self.GroupBox1.Controls.Add(self.nudPort);
  self.GroupBox1.Controls.Add(self.lblPort);
  self.GroupBox1.Controls.Add(self.lblLink);
  self.GroupBox1.Controls.Add(self.btnAction);
  self.GroupBox1.Location := new System.Drawing.Point(8, 47);
  self.GroupBox1.Name := 'GroupBox1';
  self.GroupBox1.Size := new System.Drawing.Size(528, 144);
  self.GroupBox1.TabIndex := 0;
  self.GroupBox1.TabStop := false;
  self.GroupBox1.Text := 'HttpServer';
  // 
  // lblUrl
  // 
  self.lblUrl.AutoSize := true;
  self.lblUrl.Enabled := false;
  self.lblUrl.Location := new System.Drawing.Point(98, 120);
  self.lblUrl.Name := 'lblUrl';
  self.lblUrl.Size := new System.Drawing.Size(32, 13);
  self.lblUrl.TabIndex := 8;
  self.lblUrl.Text := 'URL:';
  // 
  // nudCount
  // 
  self.nudCount.Location := new System.Drawing.Point(139, 88);
  self.nudCount.Name := 'nudCount';
  self.nudCount.Size := new System.Drawing.Size(48, 20);
  self.nudCount.TabIndex := 7;
  self.nudCount.Value := new System.Decimal(array of System.Int32([5,
      0,
      0,
      0]));
  // 
  // lblCount
  // 
  self.lblCount.AutoSize := true;
  self.lblCount.Location := new System.Drawing.Point(15, 90);
  self.lblCount.Name := 'lblCount';
  self.lblCount.Size := new System.Drawing.Size(115, 13);
  self.lblCount.TabIndex := 6;
  self.lblCount.Text := 'Listener Thread Count:';
  // 
  // lblServerName
  // 
  self.lblServerName.AutoSize := true;
  self.lblServerName.Location := new System.Drawing.Point(58, 67);
  self.lblServerName.Name := 'lblServerName';
  self.lblServerName.Size := new System.Drawing.Size(72, 13);
  self.lblServerName.TabIndex := 4;
  self.lblServerName.Text := '&Server Name:';
  // 
  // txtServerName
  // 
  self.txtServerName.Anchor := (((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Left) 
        or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.txtServerName.Location := new System.Drawing.Point(139, 64);
  self.txtServerName.Name := 'txtServerName';
  self.txtServerName.Size := new System.Drawing.Size(376, 20);
  self.txtServerName.TabIndex := 5;
  self.txtServerName.Text := 'TextBox3';
  // 
  // txtRoot
  // 
  self.txtRoot.Anchor := (((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Left) 
        or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.txtRoot.Location := new System.Drawing.Point(139, 40);
  self.txtRoot.Name := 'txtRoot';
  self.txtRoot.Size := new System.Drawing.Size(376, 20);
  self.txtRoot.TabIndex := 3;
  self.txtRoot.Text := 'TextBox2';
  // 
  // lblRoot
  // 
  self.lblRoot.AutoSize := true;
  self.lblRoot.Location := new System.Drawing.Point(75, 43);
  self.lblRoot.Name := 'lblRoot';
  self.lblRoot.Size := new System.Drawing.Size(55, 13);
  self.lblRoot.TabIndex := 2;
  self.lblRoot.Text := '&RootPath:';
  // 
  // nudPort
  // 
  self.nudPort.Location := new System.Drawing.Point(139, 16);
  self.nudPort.Name := 'nudPort';
  self.nudPort.Size := new System.Drawing.Size(48, 20);
  self.nudPort.TabIndex := 1;
  self.nudPort.Value := new System.Decimal(array of System.Int32([83,
      0,
      0,
      0]));
  self.nudPort.ValueChanged += new System.EventHandler(@self.nudPort_ValueChanged);
  // 
  // lblPort
  // 
  self.lblPort.AutoSize := true;
  self.lblPort.Location := new System.Drawing.Point(101, 18);
  self.lblPort.Name := 'lblPort';
  self.lblPort.Size := new System.Drawing.Size(29, 13);
  self.lblPort.TabIndex := 0;
  self.lblPort.Text := '&Port:';
  // 
  // lblLink
  // 
  self.lblLink.Anchor := (((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Left) 
        or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.lblLink.Enabled := false;
  self.lblLink.Location := new System.Drawing.Point(139, 120);
  self.lblLink.Name := 'lblLink';
  self.lblLink.Size := new System.Drawing.Size(168, 16);
  self.lblLink.TabIndex := 9;
  self.lblLink.TabStop := true;
  self.lblLink.Text := 'http://localhost:82/index.html';
  self.lblLink.LinkClicked += new System.Windows.Forms.LinkLabelLinkClickedEventHandler(@self.lblLink_LinkClicked);
  // 
  // btnAction
  // 
  self.btnAction.Anchor := ((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.btnAction.Location := new System.Drawing.Point(403, 115);
  self.btnAction.Name := 'btnAction';
  self.btnAction.Size := new System.Drawing.Size(112, 23);
  self.btnAction.TabIndex := 10;
  self.btnAction.Text := 'Activate Servers';
  self.btnAction.Click += new System.EventHandler(@self.btnAction_Click);
  // 
  // txtLog
  // 
  self.txtLog.Anchor := ((((System.Windows.Forms.AnchorStyles.Top or System.Windows.Forms.AnchorStyles.Bottom) 
        or System.Windows.Forms.AnchorStyles.Left) 
        or System.Windows.Forms.AnchorStyles.Right) as System.Windows.Forms.AnchorStyles);
  self.txtLog.Location := new System.Drawing.Point(8, 223);
  self.txtLog.Multiline := true;
  self.txtLog.Name := 'txtLog';
  self.txtLog.ScrollBars := System.Windows.Forms.ScrollBars.Both;
  self.txtLog.Size := new System.Drawing.Size(528, 222);
  self.txtLog.TabIndex := 2;
  self.txtLog.WordWrap := false;
  // 
  // Label2
  // 
  self.Label2.Location := new System.Drawing.Point(8, 199);
  self.Label2.Name := 'Label2';
  self.Label2.Size := new System.Drawing.Size(32, 16);
  self.Label2.TabIndex := 1;
  self.Label2.Text := 'Log';
  // 
  // pictureBox1
  // 
  self.pictureBox1.Image := (resources.GetObject('pictureBox1.Image') as System.Drawing.Image);
  self.pictureBox1.Location := new System.Drawing.Point(8, 7);
  self.pictureBox1.Name := 'pictureBox1';
  self.pictureBox1.Size := new System.Drawing.Size(120, 30);
  self.pictureBox1.TabIndex := 17;
  self.pictureBox1.TabStop := false;
  // 
  // MainForm
  // 
  self.AutoScaleBaseSize := new System.Drawing.Size(5, 13);
  self.ClientSize := new System.Drawing.Size(534, 442);
  self.Controls.Add(self.GroupBox1);
  self.Controls.Add(self.txtLog);
  self.Controls.Add(self.Label2);
  self.Controls.Add(self.pictureBox1);
  self.FormBorderStyle := System.Windows.Forms.FormBorderStyle.FixedDialog;
  self.Icon := (resources.GetObject('$this.Icon') as System.Drawing.Icon);
  self.MinimumSize := new System.Drawing.Size(550, 480);
  self.Name := 'MainForm';
  self.StartPosition := System.Windows.Forms.FormStartPosition.CenterScreen;
  self.Text := 'Internet Pack Sample Server';
  self.Load += new System.EventHandler(@self.MainForm_Load);
  self.Closed += new System.EventHandler(@self.MainForm_Closed);
  self.GroupBox1.ResumeLayout(false);
  self.GroupBox1.PerformLayout();
  (self.nudCount as System.ComponentModel.ISupportInitialize).EndInit();
  (self.nudPort as System.ComponentModel.ISupportInitialize).EndInit();
  (self.pictureBox1 as System.ComponentModel.ISupportInitialize).EndInit();
  self.ResumeLayout(false);
  self.PerformLayout();
end;
{$ENDREGION}

{$REGION Application Entry Point}
[STAThread]
class method MainForm.Main;
begin
  Application.EnableVisualStyles();

  try
    with lForm := new MainForm() do
      Application.Run(lForm);
  except
    on E: Exception do begin
      MessageBox.Show(E.Message);
    end;
  end;
end;
{$ENDREGION}

method MainForm.btnAction_Click(sender: System.Object; e: System.EventArgs);
begin
  If btnAction.Text = 'Activate Servers' Then
      ActivateServers()
    Else
      DeactivateServers();
end;

 method MainForm.ActivateServers();
 begin
    AddLog('Trying to activate servers...');
    fEchoServer := New EchoServer();
    try
        fEchoServer.Open();
        AddLog('EchoServer is active.');
    except on ex: Exception do
        AddLog('Can''t activate EchoServer. An exception occured: ' + ex.Message);
    end;


    fHttpServer := New SimpleHttpServer();
    fHttpServer.Port := Convert.ToInt32(nudPort.Value);
    fHttpServer.RootPath := txtRoot.Text;
    fHttpServer.ServerName := txtServerName.Text;
    fHttpServer.Binding.ListenerThreadCount := Convert.ToInt32(nudCount.Value);
    //fHttpServer.OnHttpRequest += new LogRequestDelegate(@Self.LogRequest);
    fHttpServer.OnHttpRequest += new OnHttpRequestHandler(LogRequest);
    fHttpServer.Open();
    AddLog(String.Format('SimpleHttpServer is active on {0} port.', fHttpServer.Port));
    SetEnable(False);
    AddLog('Servers activated.');
    btnAction.Text := 'Deactivate Servers';
  end;
   
  method MainForm.LogRequest(aSender: object; ea: OnHttpRequestArgs);
  begin
    AddLog(String.Format('Request to {0}', ea.Request.Header.RequestPath))
  end;

   
  method MainForm.SetEnable(mode: Boolean);
  begin
    lblPort.Enabled := mode;
    nudPort.Enabled := mode;
    lblServerName.Enabled := mode;
    txtServerName.Enabled := mode;
    lblRoot.Enabled := mode;
    txtRoot.Enabled := mode;
    lblUrl.Enabled := Not mode;
    lblLink.Enabled := Not mode;
  end;
  
  method MainForm.DeactivateServers();
  begin
    AddLog('Trying to deactivate servers...');    
    If Assigned(fEchoServer) Then
      fEchoServer.Close();
    AddLog('EchoServer is closed.');
    If Assigned(fHttpServer) Then
      fHttpServer.Close();
    AddLog('HttpServer is closed.');
    SetEnable(True);
    AddLog('Servers is deactivated');
    btnAction.Text := 'Activate Servers';
  end;

method MainForm.lblLink_LinkClicked(sender: System.Object; e: System.Windows.Forms.LinkLabelLinkClickedEventArgs);
begin
  if (File.Exists(fHttpServer.RootPath + '\index.html')) then
    System.Diagnostics.Process.Start(lblLink.Text)
  else 
    MessageBox.Show(fHttpServer.RootPath + '\index.html can not be opened, because it does not exists.', 'Warning', 
                        MessageBoxButtons.OK, MessageBoxIcon.Warning);
end;

method MainForm.MainForm_Load(sender: System.Object; e: System.EventArgs);
begin
  txtRoot.Text := Path.Combine(Path.GetDirectoryName(Self.GetType().Assembly.Location), 'HttpRoot');
  txtServerName.Text := 'Internet Pack HTTP Server';
  nudPort.Value := 82;
  nudCount.Value := 5;
end;

method MainForm.AddLog(line: String);
begin
  txtLog.Invoke(method begin
      txtLog.AppendText(Environment.NewLine + String.Format('{0}: {1}', DateTime.Now.ToLongTimeString(), line));
  end);
end;

method MainForm.MainForm_Closed(sender: System.Object; e: System.EventArgs);
begin
  DeactivateServers();
end;

method MainForm.nudPort_ValueChanged(sender: System.Object; e: System.EventArgs);
begin
  lblLink.Text := String.Format('http://localhost:{0}/index.html', nudPort.Value);
end;

end.

File: Samples/C#/Virtual FTP/MainForm.cs
using System;
using System.IO;
using System.Windows.Forms;
using RemObjects.InternetPack.Ftp.VirtualFtp;

namespace VirtualFtp
{
	public class MainForm : System.Windows.Forms.Form
	{
		const int port = 4444;
		private System.Windows.Forms.Button btnStartStop;
		private System.Windows.Forms.GroupBox GroupBox1;
		private System.Windows.Forms.TextBox txtLog;
		private System.Windows.Forms.LinkLabel llShortcut;
		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.PictureBox pictureBox1;

		private VirtualFolder fRootFolder;
		private IFtpUserManager fUserManager;
		private VirtualFtpServer fFtpServer;

		#region Boilerplate stuff
		private System.ComponentModel.Container components = null;

		public MainForm()
		{
			InitializeComponent();
		}

		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose(disposing);

		}

		private void InitializeComponent()
		{
			System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(MainForm));
			this.btnStartStop = new System.Windows.Forms.Button();
			this.GroupBox1 = new System.Windows.Forms.GroupBox();
			this.txtLog = new System.Windows.Forms.TextBox();
			this.llShortcut = new System.Windows.Forms.LinkLabel();
			this.label1 = new System.Windows.Forms.Label();
			this.pictureBox1 = new System.Windows.Forms.PictureBox();
			this.GroupBox1.SuspendLayout();
			((System.ComponentModel.ISupportInitialize)(this.pictureBox1)).BeginInit();
			this.SuspendLayout();
			// 
			// btnStartStop
			// 
			this.btnStartStop.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
			this.btnStartStop.Location = new System.Drawing.Point(231, 8);
			this.btnStartStop.Name = "btnStartStop";
			this.btnStartStop.Size = new System.Drawing.Size(75, 23);
			this.btnStartStop.TabIndex = 18;
			this.btnStartStop.Text = "Start";
			this.btnStartStop.Click += new System.EventHandler(this.btnStartStop_Click);
			// 
			// GroupBox1
			// 
			this.GroupBox1.Anchor = ((System.Windows.Forms.AnchorStyles)((((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom)
						| System.Windows.Forms.AnchorStyles.Left)
						| System.Windows.Forms.AnchorStyles.Right)));
			this.GroupBox1.Controls.Add(this.txtLog);
			this.GroupBox1.Location = new System.Drawing.Point(8, 103);
			this.GroupBox1.Name = "GroupBox1";
			this.GroupBox1.Size = new System.Drawing.Size(298, 132);
			this.GroupBox1.TabIndex = 17;
			this.GroupBox1.TabStop = false;
			this.GroupBox1.Text = "Log";
			// 
			// txtLog
			// 
			this.txtLog.Anchor = ((System.Windows.Forms.AnchorStyles)((((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom)
						| System.Windows.Forms.AnchorStyles.Left)
						| System.Windows.Forms.AnchorStyles.Right)));
			this.txtLog.Location = new System.Drawing.Point(8, 16);
			this.txtLog.Multiline = true;
			this.txtLog.Name = "txtLog";
			this.txtLog.ScrollBars = System.Windows.Forms.ScrollBars.Both;
			this.txtLog.Size = new System.Drawing.Size(282, 108);
			this.txtLog.TabIndex = 6;
			this.txtLog.WordWrap = false;
			// 
			// llShortcut
			// 
			this.llShortcut.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left)
						| System.Windows.Forms.AnchorStyles.Right)));
			this.llShortcut.Location = new System.Drawing.Point(4, 62);
			this.llShortcut.Name = "llShortcut";
			this.llShortcut.Size = new System.Drawing.Size(300, 33);
			this.llShortcut.TabIndex = 15;
			this.llShortcut.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
			this.llShortcut.LinkClicked += new System.Windows.Forms.LinkLabelLinkClickedEventHandler(this.llShortcut_LinkClicked);
			// 
			// label1
			// 
			this.label1.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left)
						| System.Windows.Forms.AnchorStyles.Right)));
			this.label1.Location = new System.Drawing.Point(3, 38);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(309, 24);
			this.label1.TabIndex = 16;
			this.label1.Text = "In order to login on ftp please use login: test; password: test.";
			this.label1.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
			// 
			// pictureBox1
			// 
			this.pictureBox1.Image = ((System.Drawing.Image)(resources.GetObject("pictureBox1.Image")));
			this.pictureBox1.Location = new System.Drawing.Point(8, 7);
			this.pictureBox1.Name = "pictureBox1";
			this.pictureBox1.Size = new System.Drawing.Size(120, 30);
			this.pictureBox1.SizeMode = System.Windows.Forms.PictureBoxSizeMode.AutoSize;
			this.pictureBox1.TabIndex = 14;
			this.pictureBox1.TabStop = false;
			// 
			// MainForm
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(314, 242);
			this.Controls.Add(this.btnStartStop);
			this.Controls.Add(this.GroupBox1);
			this.Controls.Add(this.llShortcut);
			this.Controls.Add(this.label1);
			this.Controls.Add(this.pictureBox1);
			this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
			this.MinimumSize = new System.Drawing.Size(330, 280);
			this.Name = "MainForm";
			this.Tag = "";
			this.Text = "Virtual FTP";
			this.Closed += new System.EventHandler(this.MainForm_Closed);
			this.GroupBox1.ResumeLayout(false);
			this.GroupBox1.PerformLayout();
			((System.ComponentModel.ISupportInitialize)(this.pictureBox1)).EndInit();
			this.ResumeLayout(false);
			this.PerformLayout();

		}

		[STAThread]
		static void Main()
		{
			Application.EnableVisualStyles();
			Application.Run(new MainForm());
		}
		#endregion

		private void btnStartStop_Click(object sender, System.EventArgs e)
		{
			if (btnStartStop.Text == "Start")
			{
				addToLog("Starting Virtual FTP on " + port.ToString() + " port...");
				StartServer();
				llShortcut.Text = String.Format("ftp://127.0.0.1:{0}/", port);
				addToLog("Virtual FTP is running under " + Environment.OSVersion);
				btnStartStop.Text = "Stop";
			}
			else
			{
				addToLog("Shutting down Virtual FTP ...");
				StopServer();
				llShortcut.Text = "";
				addToLog("Virtual FTP is stopped.");
				btnStartStop.Text = "Start";
			}
		}

		private void llShortcut_LinkClicked(object sender, System.Windows.Forms.LinkLabelLinkClickedEventArgs e)
		{
			if (llShortcut.Text != "") System.Diagnostics.Process.Start(llShortcut.Text);
		}

		private void MainForm_Closed(object sender, System.EventArgs e)
		{
			StopServer();
		}

		private void StartServer()
		{
			string lDiskFolder = Path.GetDirectoryName(typeof(MainForm).Assembly.Location) + @"\FtpRoot";

			fRootFolder = new VirtualFolder(null, "[ROOT]");
			fRootFolder.Add(new VirtualFolder(null, "virtual"));
			fRootFolder.Add(new DiscFolder(null, "drive-c", @"c:\"));
			fRootFolder.Add(new DiscFolder(null, "disc", lDiskFolder));
			fRootFolder.Add(new EmptyFile(null, "=== Welcome to the FTP ==="));

			fUserManager = new UserManager();
			((UserManager)fUserManager).AddUser("test", "test");

			fFtpServer = new VirtualFtpServer();
			fFtpServer.Port = port;
			fFtpServer.Timeout = 60 * 1000; /* 1 minute */
			if (this.fFtpServer.BindingV4 != null)
				this.fFtpServer.BindingV4.ListenerThreadCount = 10;
			fFtpServer.RootFolder = fRootFolder;
			fFtpServer.UserManager = fUserManager;
			fFtpServer.ServerName = "VirtualFTP Sample - powered by RemObjects Internet Pack for .NET";

			fFtpServer.Open();

			addToLog("VirtualFTP 0.3 BETA - started up");
		}

		private void StopServer()
		{
			if (fFtpServer != null)
				fFtpServer.Close();
		}

		private void addToLog(String line)
		{
			txtLog.Text = txtLog.Text +
				System.DateTime.Now.ToLongTimeString() +
				": " +
				line +
				Environment.NewLine;
		}
	}
}